### PR TITLE
Adds ActionIndex and Harmonize use of action-masks

### PR DESCRIPTION
**Pull request by julienroyd**
Masks were used somewhat ad-hoc here and there. Here we centralise masking in GraphActionCategorical(), specifically:

- renamed masks in GraphActionCategorical to action_mask (to avoid confusion with done-masking, molecule validity masking, goal-conditioning masking, etc.)
- made logits and masks private attributes of GraphActionCategorical
- remove _mask() application from GraphTransformerFragEnvelopeQL and GraphTransformerGFN
- removed masking from GraphSampler (masks applied in property setters of GraphActionCategorical._logits)

We also replaced the action-tuples `Tuple[int, int, int]` action type, row and column by a named-tuple `ActionIndex` to the code and typing more intuitive and readable.

### CODE REVIEW

**Comment by julienroyd** [comment]
Can reproduce previous results with goal-conditioning and 4 objectives:
image

File: src/gflownet/envs/graph_building_env.py [hunk 1]
```diff
@@ -696,35 +749,18 @@ def sample(self) -> List[Tuple[int, int, int]]:
         # mutually exclusive).
 
         # Uniform noise
-        u = [torch.rand(i.shape, device=self.dev) for i in self.logits]
+        u = [torch.rand(i.shape, device=self.dev) for i in self._masked_logits]
         # Gumbel noise
-        gumbel = [logit - (-noise.log()).log() for logit, noise in zip(self.logits, u)]
-
-        if self.masks is not None:
-            gumbel_safe = [
```
Comment (bengioe) [inline src/gflownet/envs/graph_building_env.py:704]: This is removed because we're now using `-inf`?

File: src/gflownet/algo/graph_sampling.py [hunk 2]
```diff
@@ -259,21 +256,17 @@ def not_done(lst):
             else:
                 gbatch = self.ctx.collate(torch_graphs)
                 action_types = self.ctx.bck_action_type_order
-                masks = [getattr(gbatch, i.mask_name) for i in action_types]
+                action_masks = [self.ctx.action_type_to_mask(t, gbatch, assert_mask_exists=True) for t in action_types]
```
Comment (bengioe) [inline src/gflownet/algo/graph_sampling.py]: `action_type_to_mask` seems to be missing for `mol_building_env` and `seq_building_env`?

**Review by bengioe — APPROVED** [review]
LGTM, just two questions.

File: src/gflownet/envs/graph_building_env.py (hunk 1 above)
Comment (julienroyd) [reply src/gflownet/envs/graph_building_env.py:704]: Yes. I talked with Sobhan and he told me that he added that in because the multiplicative masking would not work if the model was unstable and outputting extremely small values for the logits. Masking with -inf should fix the issue.

File: src/gflownet/algo/graph_sampling.py (hunk 2 above)
Comment (bengioe) [reply src/gflownet/algo/graph_sampling.py]: Actually, I'm not sure this needs to be a context method (nor a graph model method), it could just be a function somewhere?

Comment (julienroyd) [reply src/gflownet/algo/graph_sampling.py]: Yeah I agree, I wasn't sure where to put that. I can simply put it as a general function in `graph_building_env.py` for now?

Comment (bengioe) [reply src/gflownet/algo/graph_sampling.py]: 👍  works for me.
